docs/abi.rst

```diff
@@ -0,0 +1,919 @@
+
+The :any:`BareCallActions` class is used to define a bare app call handler for on completion actions. Each bare app call handler must be an instance of the :any:`OnCompleteAction` class.
+
+The :any:`OnCompleteAction` class is responsible for holding the actual code for the bare app call handler (an instance of either :code:`Expr` or a subroutine that takes no args and returns nothing) as well as a :any:`CallConfig` option that indicates whether the action is able to be called during a creation app call, a non-creation app call, or either.
+
+All the bare app calls that an application wishes to support must be provided to the :any:`Router.__init__` method.
+
+A brief example is below:
+
+.. code-block:: python
+
+    from pyteal import *
+
+    @Subroutine(TealType.none)
+    def opt_in_handler() -> Expr:
+        return App.localPut(Txn.sender(), Bytes("opted_in_round"), Global.round())
+
+
+    @Subroutine(TealType.none)
+    def assert_sender_is_creator() -> Expr:
+        return Assert(Txn.sender() == Global.creator_address())
+
+
+    router = Router(
+        name="ExampleApp",
+        bare_calls=BareCallActions(
+            # Allow app creation with a no-op action
+            no_op=OnCompleteAction(
+                action=Approve(), call_config=CallConfig.CREATE
+            ),
+
+            # Register the `opt_in_handler` to be called during opt in.
+            #
+            # Since we use `CallConfig.ALL`, this is also a valid way to create this app
+            # (if the creator wishes to immediately opt in).
+            opt_in=OnCompleteAction(
+                action=opt_in_handler, call_config=CallConfig.ALL
+            ),
+
+            # Allow anyone who opted in to close out from the app.
+            close_out=OnCompleteAction(
+                action=Approve(), call_config=CallConfig.CALL
+            ),
+
+            # Only approve update and delete operations if `assert_sender_is_creator` succeeds.
+            update_application=OnCompleteAction(
+                action=assert_sender_is_creator, call_config=CallConfig.CALL
+            ),
+            delete_application=OnCompleteAction(
+                action=assert_sender_is_creator, call_config=CallConfig.CALL
+            ),
+        ),
+    )
+
+.. note::
+    When deciding which :any:`CallConfig` value is appropriate for a bare app call or method, consider the question, should it be valid for someone to create my app with this operation? Most of the time the answer will be no, in which case :any:`CallConfig.CALL` should be used.
+
+Registering Methods
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. warning::
+    A challenge for any system exposed to the public, including smart contracts, is input validation. At the moment, PyTeal's :any:`Router` class only performs a limited set of input validation. The largest hole in this validation is compound types, i.e. :any:`abi.StaticArray`, :any:`abi.Address`, :any:`abi.DynamicArray`, and :any:`abi.String`.
```

Comment:
@jasonpaulos Optional:  I think the _warning_ provides good content, but can be more concise + have fewer long sentences.  If you prefer as is, feel welcomed to resolve.

Here's a proposal for slimming down the warning while hopefully retaining its intent.  If you don't like the proposal, feel welcomed to disregard it.

:any:`Router` does _not_ validate compound types (:any:`abi.StaticArray`, :any:`abi.Address`, :any:`abi.DynamicArray`, and :any:`abi.String`).  **We strongly recommend** methods immediately access and validate compound type parameters _before_ persisting arguments for later transactions.

Notes:
* The recommendation applies to recursively defined compound types.
* :any:`abi.Address` is _not_ guaranteed to have exactly 32 bytes.  To defend against unintended behavior, manually verify the length is 32 bytes.

-----

Aside:  Additionally and also optionally, I'd consider adding _compound types_ to an existing section header.  
* That way the warning can refer to the section defining compound types rather than re-enumerating the types. 
* Candidate section:  _Getting Values at Indexes_ can be named something like _Getting Values at Indexes - Compound types_

Reply:
Thanks for the suggestions, I've adopted a variation on this